marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Api.hs

```diff
@@ -0,0 +1,363 @@
+{-# LANGUAGE EmptyDataDeriving #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Language.Marlowe.Runtime.Transaction.Api
+  where
+
+import Cardano.Api
+  ( AsType(..)
+  , ScriptDataSupportedInEra(..)
+  , SerialiseAsRawBytes(serialiseToRawBytes)
+  , Tx
+  , TxBody
+  , deserialiseFromCBOR
+  , deserialiseFromRawBytes
+  , serialiseToCBOR
+  )
+import Cardano.Api.Shelley (StakeCredential(..))
+import qualified Data.Aeson as Aeson
+import Data.Binary (Binary, Get, get, getWord8, put)
+import Data.Binary.Put (Put, putWord8)
+import Data.ByteString (ByteString)
+import Data.Map (Map)
+import Data.Set (Set)
+import Data.Some (Some(..))
+import Data.Time (UTCTime)
+import Data.Type.Equality (type (:~:)(Refl))
+import Data.Void (Void, absurd)
+import GHC.Generics (Generic)
+import Language.Marlowe.Runtime.ChainSync.Api (Address, BlockHeader, TokenName, TxId, getUTCTime, putUTCTime)
+import Language.Marlowe.Runtime.Core.Api
+import Language.Marlowe.Runtime.Transaction.Constraints (UnsolvableConstraintsError)
+import Network.Protocol.Job.Types (Command(..), SomeTag(..))
+
+-- | The low-level runtime API for building and submitting transactions.
+data MarloweTxCommand status err result where
+  -- | Construct a transaction that starts a new Marlowe contract. The
+  -- resulting, unsigned transaction can be signed via the cardano API or a
+  -- wallet provider. When signed, the 'Submit' command can be used to submit
+  -- the transaction to the attached Cardano node.
+  Create
+    :: ScriptDataSupportedInEra era
+    -> Maybe StakeCredential
+    -- ^ A reference to the stake address to use for script addresses.
+    -> MarloweVersion v
+    -- ^ The Marlowe version to use
+    -> WalletAddresses
+    -- ^ The wallet addresses to use when constructing the transaction
+    -> Map TokenName Address
+    -- ^ The initial distribution of role tokens
+    -> Map Int Aeson.Value
+    -- ^ Optional metadata to attach to the transaction
+    -> Contract v
+    -- ^ The contract to run
+    -> MarloweTxCommand Void CreateError
+        ( ContractId -- The ID of the contract (tx output that carries the datum)
+        , TxBody era -- The unsigned tx body, to be signed by a wallet.
+        )
+
+  -- | Construct a transaction that advances an active Marlowe contract by
+  -- applying a sequence of inputs. The resulting, unsigned transaction can be
+  -- signed via the cardano API or a wallet provider. When signed, the 'Submit'
+  -- command can be used to submit the transaction to the attached Cardano node.
+  ApplyInputs
+    :: ScriptDataSupportedInEra era
+    -> MarloweVersion v
+    -- ^ The Marlowe version to use
+    -> WalletAddresses
+    -- ^ The wallet addresses to use when constructing the transaction
+    -> ContractId
+    -- ^ The ID of the contract to apply the inputs to.
+    -> Maybe UTCTime
+    -- ^ The "invalid before" bound of the validity interval. If omitted, this
```

Comment:
Possible, but we'd need the current slot number of the node, etc.

Reply:
That can easilly be obtained from `chainseekd` - in fact, I would go so far as to say this should be the _preferred_ way to do it, as it avoids that mess we ran into when trying to do it from Marlowe Run.